wave_utils/src/pose_cov_comp.cpp

```diff
@@ -0,0 +1,514 @@
+ |                            O/_____/_______/____\O                          |
+ |                            /____________________\                          |
+ |                           /    (#UNIVERSITY#)    \                         |
+ |                           |[**](#OFWATERLOO#)[**]|                         |
+ |                           \______________________/                         |
+ |                            |_""__|_,----,_|__""_|                          |
+ |                            ! !                ! !                          |
+ |                            '-'                '-'                          |
+ |       __    _   _  _____  ___  __  _  ___  _    _  ___  ___   ____  ____   |
+ |      /  \  | | | ||_   _|/ _ \|  \| |/ _ \| \  / |/ _ \/ _ \ /     |       |
+ |     / /\ \ | |_| |  | |  ||_||| |\  |||_|||  \/  |||_||||_|| \===\ |====   |
+ |    /_/  \_\|_____|  |_|  \___/|_| \_|\___/|_|\/|_|\___/\___/ ____/ |____   |
+ |                                                                            |
+ ******************************************************************************
+ * ############################################################################
+ *
+ * File: ref_ekf.cpp
+ * Desc: File containing the impl. for pose composition with uncertainty
+ * Auth: Chunshang Li and Jordan Hu
+ *
+ * ############################################################################
+ */
+
+#include "wave/utils/pose_cov_comp.hpp"
+#include <Eigen/Core>
+#include <Eigen/Dense>
+
+namespace wave {
+
+PoseWithCovariance::PoseWithCovariance() {
+    this->position.setZero();
+    this->rotation_matrix.setIdentity();
+    this->covariance.setZero();
+}
+
+PoseWithCovariance::PoseWithCovariance(Vector6 &p, Matrix6x6 &cov) {
+    Vector3 ypr;
+
+    ypr << p[5], p[4], p[3];
+
+    this->position = p.block<3, 1>(0, 0);
+    this->rotation_matrix = yprToRotMatrix(ypr);
+    this->covariance = cov;
+}
+
+PoseWithCovariance::PoseWithCovariance(Vector3 &p,
+                                       Matrix3x3 &r,
+                                       Matrix6x6 &cov) {
+    this->position = p;
+    this->rotation_matrix = r;
+    this->covariance = cov;
+}
+
+
+Vector3 PoseWithCovariance::getPosition() const {
+    return this->position;
+}
+
+Vector3 PoseWithCovariance::getYPR() const {
+    Vector3 ypr = rotMatrixToYPR(this->rotation_matrix);
+    return ypr;
+}
+
+Vector4 PoseWithCovariance::getQuaternion() const {
+    Vector4 q = rotMatrixToQuat(this->rotation_matrix);
+
+    q = normalizeQuat(q);
+
+    return q;
+}
+
+Vector7 PoseWithCovariance::getPoseQuaternion() const {
+    Vector7 pose;
+    Vector4 q = this->getQuaternion();
+
+    pose << this->position, q;
+
+    return pose;
+}
+
+Eigen::Affine3d PoseWithCovariance::getTransformMatrix() const {
+    Eigen::Affine3d T_m;
+
+    T_m.translation() = this->position;
+    T_m.linear() = this->rotation_matrix;
```

Comment:
I thought `Eigen::Affine3d` also has a `rotation()` method - but I don't quite remember if that's assignable. 

https://stackoverflow.com/questions/35416880/what-does-transformlinear-return-in-the-eigen-library

Interesting, that's good to know. This is fine then.